### DISCUSSION TITLE
Set Mouse Position or Lock Mouse

### DISCUSSION THREAD

**Discussion by Haltroy** [AvaloniaUI/Avalonia#12536] [Q&A] [answered]
Hi,
Is there any way to set the position of mouse (or pointer) from `PointerEnter` (or similar) event?

**Comment by thevortexcloud** [comment]
It's fairly simple if you are willing to dip into native code. EG Windows has `MOUSEINPUT` and X11 has `XWarpPointer`. I am not sure if Avalonia can do this itself though. There seems to be a DllImport for `XWarpPointer` in Avalonia but I don't think it's being used anywhere. Avalonia in general itself considers everything to be a generic pointing device, which includes things like touch which do not have a concept of a cursor. As such I doubt it's being exposed.

**Reply by Haltroy** [reply]
So it's a no for someone like me who can't dip into native code?
Well is there any library or code that could help with my issue where a cross-platform app uses `MOUSEINPUT` on Windows and `XWarpPointer` on X11 stuff? I am really terrible at native code stuff.

I will try to find a way to see if I can implement it and if I manage to do it, I would probably create a PR here that would implement a function to reposition pointer to a point/control/etc..

**Reply by thevortexcloud** [reply]
As far as I am aware, there is not. But I generally just write my own for this sort of stuff (at least on Windows). It's not a hard thing to abstract generally. That said you would likely also need to support MacOS (as well as any other platform that allows setting cursor position) if you wanted to contribute this back to Avalonia (although I am not sure such a request would even be accepted due to how generic the input system tries to be. But the Avalonia devs can probably comment on that better than I can). There are examples here for how to do it for Windows though.

https://stackoverflow.com/questions/1392199/how-to-set-mouse-cursor-position-to-a-specified-point-on-screen-in-c

X11 documentation/examples is generally far harder to find unless you are willing to look at C code/read man pages. But as I said, it looks like Avalonia already has the signature to use `XWarpPointer` in it's X11 code. You can probably figure out how to piece it together using the existing X11 code if you want to modify Avalonia itself. That said you could also use uinput instead of doing it via X11. It should be far more portable since you are more or less creating a virtual mouse you have full control over. Which means it should work properly on X11 and Wayland. The documentation for that is here (I believe you would need to use `EV_ABS` instead of `EV_REL`?)

https://kernel.org/doc/html/v6.0/input/uinput.html
https://www.kernel.org/doc/Documentation/input/event-codes.txt

But uinput is something I have only briefly messed with in the past. 

Personally I find it easier to just write a native library that wraps most of the more complicated stuff for me, which makes the C# interop significantly simpler as it means there is less methods/structs and what not to worry about. But that's not a requirement. Rust is also much easier to get up and running for this than C if you do decide to do that (assuming Rust bindings exist for what ever you try to do. Which they do for X11 and uinput).

**Comment by timunie** [comment]
I'm not sure if that is something we want to have implemented by default. While there may be some rare cases where this is totally valid thing to do, I *hate* if any program thinks it should move my pointer for me. I'd click in a wrong position all the time 😆 

Happy coding

**Reply by thevortexcloud** [reply]
Funnily enough this is more or less what the documentation for `XWarpPointer` says.

> There is seldom any reason for calling this function. The pointer should normally be left to the user.

**Reply by Haltroy** [reply]
Yes I know but I was trying to make a simple game where player would solve mazes by using their mouse and reset the mouse position to a point when they hit an obstacle or teleport the mouse to another point on portals etc.

I actually made the game in WinForms back in the day where everything was too simple.

**Reply by maxkatz6** [reply]
We might add pointer lock API in the future, but we won't add "set mouse position" apis.
Something similar to HTML5 function https://developer.mozilla.org/en-US/docs/Web/API/Pointer_Lock_API

**Reply by Haltroy** [reply]
"Pointer Lock", I can work with that.

Postponing my game idea until that API is implemented.

**Reply by timunie** [reply]
Don't expect this to happen too early.

**Reply by thevortexcloud** [reply]
This does however mean a contribution adding it would be more likely to be accepted if you wanted to add that yourself.

**Reply by Haltroy** [reply]
> Don't expect this to happen too early.

No worries, I can wait.

> This does however mean a contribution adding it would be more likely to be accepted if you wanted to add that yourself.

Yeah, I will make a PR if I figure it out but that might take some time.

**Comment by Haltroy** [comment]
I realized that I don't really need this function that much and just tell the player to hover on that start position for now until that `lock mouse` feature is implemented (by either me or someone else). For now;
 - There's no in direct way to set mouse position in Avalonia.
 - There are native ways to set the mouse position for Windows and X11 (Linux)
   - X11's `XWarpPointer` asks developers to not use this to not annoy user. [Reference](https://tronche.com/gui/x/xlib/input/XWarpPointer.html#:~:text=There%20is%20seldom%20any%20reason%20for%20calling%20this%20function.%20The%20pointer%20should%20normally%20be%20left%20to%20the%20user.)
   - Just use WinForm's `System.Windows.Forms.Cursor.Position` for Windows. (`System.Windows.Forms.Cursor.Position = new Point(x,y);`)
   - No idea on Mac (they no longer use X11 and I don't own a mac), sorry.
   
 Also, nothing about "annoying user" makes sense in my case because this was for my hobby game project where players would solve a puzzle using their mouse and due to the nature of this game the player will obviously get annoyed by it and there would be a way to escape the trap by just pressing a button (ex. ESC like any game) which will bring a pause menu.

**Comment by Haltroy** [comment, marked as answer]
After working on some solution, I decided to just make player drag a rectangle to end instead of using the mouse point only. However, I did get the "warping mouse" part done partially using SDL. It mostly works, except in Wayland sessions. 

# WARNING: I DONT RECOMMEND WARPING MOUSE AS XORG DOCUMENT NOTES, IT IS REALLY ANNOYING THE USER. ONLY USE IT IF YOU REALLY NEED TO

Use this to check if running on a Wayland session (mostly works on most cases):
```csharp
RuntimeInformation.IsOSPlatform(OSPlatform.Linux) && System.Environment.GetEnvironmentVariable("XDG_SESSION_TYPE").ToLowerInvariant() == "wayland"
```

To warp the mouse (using [SDL2](https://github.com/flibitijibibo/SDL2-CS)):

```csharp
public static void WarpMouse(int x, int y)
    {
        SDL.SDL_Init(SDL.SDL_INIT_VIDEO);

        IntPtr window = SDL.SDL_CreateWindow("Move Mouse", 0, 0, 0, 0, SDL.SDL_WindowFlags.SDL_WINDOW_SHOWN);

        if (window != IntPtr.Zero)
        {

            SDL.SDL_RenderPresent(SDL.SDL_GetRenderer(window));


            SDL.SDL_WarpMouseInWindow(window, x, y);


            SDL.SDL_DestroyWindow(window);
            SDL.SDL_Quit();

        }
    }
```
